OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,980 @@
+				return;
+			var mcvType = Info.McvTypes.Random(world.LocalRandom);
+
+			// Make sure we only request one MCV at a time.
+			if (unitBuilder.RequestedProductionCount(bot, mcvType) <= 0)
+				unitBuilder.RequestUnitProduction(bot, mcvType);
+		}
+
+		void DeployMcvs(IBot bot, bool chooseLocation)
+		{
+			var newMCVs = world.ActorsHavingTrait<Transforms>()
+				.Where(a => a.Owner == player && a.IsIdle && Info.McvTypes.Contains(a.Info.Name));
+
+			foreach (var mcv in newMCVs)
+				DeployMcv(bot, mcv, chooseLocation);
+		}
+
+		void UnDeployConyard(IBot bot)
+		{
+			/*
+			 * Current BaseBuilderBotModule has a very weak control of BO (Build Orders), produce randomly and cannot cooperate well,
+			 * which makes expanding with only 1 MCV performs terrible: bot's necessary base building always get interrupted and spend too many time on road.
+			 *
+			 * Therefore, the considerations of undeploy and move MCV are simple and conservative, at least for now.
+			 *
+			 * TODO: This should be tied deeply with BaseBuilderBotModule in order to keep better control on having only 1 MCV:
+			 * undeploy should affect BaseBuilderBotModule's BO, while BaseBuilderBotModule has some control on MCV's expansion,
+			 * in order to make expansion worth the time and risk.
+			 */
+			if (firstUndeploy)
+			{
+				/*
+				 * On the first undeploy, we only consider if bot player encounter the early game rush, when bot player
+				 * has less than 2 production buildings (barrack, war factory, etc) or no refinery, MCV won't move and leave everything behind.
+				 */
+				if (world.ActorsHavingTrait<Building>().Count(a => a.Owner == player && Info.ProductionTypes.Contains(a.Info.Name)) >= 2
+					&& world.ActorsHavingTrait<Refinery>().Any(a => a.Owner == player && Info.RefineryTypes.Contains(a.Info.Name)))
+				{
+					var idleconyards = constructionYards.Actors.Where(a => !a.IsDead).ToList();
+
+					if (idleconyards.Count > 0)
+					{
+						bot.QueueOrder(new Order("DeployTransform", idleconyards[0], true));
+						allowfallback = false;
+					}
+				}
+
+				firstUndeploy = false;
+				moveConyardInterval = world.LocalRandom.Next(Info.MoveConyardTick, Info.MoveConyardTick * 2);
+			}
+			else
+			{
+				/*
+				 * On regular undeploy, we only consider that the bot always keep a deployed conyard (so this only works with 2 or more conyards at field),
+				 * and moving MCV should never break the production of refinery.
+				 *
+				 * Because breaking the production of refinery will cause a dead loop of:
+				 * MCV deploys at expansion location -> MCV is producing refinery -> MCV unpacks and interrupts production -> MCV find no refinery here, see this as expansion location and deploy
+				 */
+				var conyards = constructionYards.Actors
+					.Where(a => !a.IsDead).OrderBy(a => a.ActorID)
+					.ToList();
+
+				if (conyards.Count > 1)
+				{
+					var movableMCV = conyards.FirstOrDefault(a => !a.TraitsImplementing<ProductionQueue>()
```

Comment:
```suggestion
					// We don't want to interrupt refinery production, otherwise it may cause a dead loop of deploy/undeploy.
					var movableMCV = conyards.FirstOrDefault(a => !a.TraitsImplementing<ProductionQueue>()
```

maybe move the comment here

Reply:
fixed